### PR TITLE
Fix to #12358 - The use of the ToListAsync method leads to missing data in version 2.1.0 of Microsoft.AspNetCore.All

### PR DESCRIPTION
**Pull request by maumar**
Problem was in the implementation of TaskLiftingExpressionVisitor and how it interacts with correlated collection optimization. When we perform task lift we do so recursively if we haven't encountered a CancellationToken parameter yet (essentially fishing for it, so that we know what to use).
This becomes a problem if correlated collection optimization async is inside another async method (like Task.FromResult). We ended up recursively lifting the task twice, which resulted in the same correlate call executed twice, which is a problem because both calls share reference to the same reader.

Fix is to fetch CancelationToken parameter as a separate step and don't recursively lift the tasks. It can still deadlock potentially, but will now return correct results.

### CODE REVIEW

**Comment by maumar** [comment]
we already have tests: Correlated_collections_nested_inner_subquery_references_outer_qsre_one_level_up and ToArray_on_nav_subquery_in_projection_nested

File: src/EFCore/Query/ExpressionVisitors/Internal/TaskLiftingExpressionVisitor.cs [hunk 1]
```diff
@@ -39,8 +41,9 @@ public class TaskLiftingExpressionVisitor : RelinqExpressionVisitor
         /// </summary>
         public virtual Expression LiftTasks([NotNull] Expression expression)
         {
-            var newExpression = Visit(expression);
+            _cancelationTokenFinder = new CancelationTokenFindingExpressionVisitor();
```
Comment (smitpatel) [inline src/EFCore/Query/ExpressionVisitors/Internal/TaskLiftingExpressionVisitor.cs]: `_cancellation`

File: src/EFCore/Query/ExpressionVisitors/Internal/TaskLiftingExpressionVisitor.cs [hunk 2]
```diff
@@ -27,6 +27,8 @@ public class TaskLiftingExpressionVisitor : RelinqExpressionVisitor
 
         private readonly List<Expression> _taskExpressions = new List<Expression>();
 
+        private CancelationTokenFindingExpressionVisitor _cancelationTokenFinder;
```
Comment (AndriySvyryd) [inline src/EFCore/Query/ExpressionVisitors/Internal/TaskLiftingExpressionVisitor.cs]: Remove `_dummyCancellationToken` and review when `CancellationToken` should be looked for at all

**Comment by maumar** [comment]
merged 3d2ba391cc1d8d6cc7b0a67ced4539067ad07a64

**Pull request closed without merging**